### PR TITLE
don't crash on block production error; build lookahead proposal from current head

### PR DESCRIPTION
**Pull request by tersec**
This also removes the nested `withState` usage, which was not optimally clear.

Current `stable` does not crash on block production errors -- this was introduced with the bid comparison logic.

### CODE REVIEW

File: beacon_chain/consensus_object_pools/consensus_manager.nim [hunk 1]
```diff
@@ -334,14 +334,18 @@ proc runProposalForkchoiceUpdated*(
...
+      if self.dag.headState.kind >= ConsensusFork.Capella:
         # Within an epoch, so long as there's no block, the withdrawals also
         # remain unchanged. Balances change at epoch boundaries, however, so
         # if and only if the proposal slot is the first slot of an epoch the
         # beacon node must transition epochs to compute correct balances.
         if nextWallSlot.is_epoch:
           var cache: StateCache
+
+          # getProposalState builds off clearanceState
+          assign(self.dag.clearanceState, self.dag.headState)
```
Comment (arnetheduck) [inline beacon_chain/consensus_object_pools/consensus_manager.nim]: this assignment looks dubious - `getProposalState` already should get the proposal state from the "best" source which is "primed" ahead of each slot - this assignment thus should be redundant in most cases and might worsen performance in others

Comment (arnetheduck) [reply beacon_chain/consensus_object_pools/consensus_manager.nim]: ie there exists a small opportunity to optimise this flow to avoid a copy in some cases, but should be done inside `getProposalState` - in particular, we want to run the slot transition right before the slot of the actual proposal and avoid as many state assignments as possible

Comment (tersec) [reply beacon_chain/consensus_object_pools/consensus_manager.nim]: > we want to run the slot transition right before the slot of the actual proposal and avoid as many state assignments as possible

Sure, then can't pre-prepare correct withdrawals when proposal slot is epoch-aligned. The alternate option is to just not send early-fcUs for epoch-aligned proposal slots in Capella and later, and leave that logic purely in the fallback flow in `validator_duties`. I'm fine with that too.

Comment (tersec) [reply beacon_chain/consensus_object_pools/consensus_manager.nim]: > this assignment looks dubious - `getProposalState` already should get the proposal state from the "best" source which is "primed" ahead of each slot - this assignment thus should be redundant in most cases and might worsen performance in others

The early fcU angle is why this doesn't happen.

Comment (arnetheduck) [reply beacon_chain/consensus_object_pools/consensus_manager.nim]: I'm missing some flow here then: `getProposalState` should already look at all available "cached" states and select the one that most closely matches the desired slot combination (ie requires the least amount of work) - ie one of the options it should be looking at is the head state as assigned here, based on its "desired" state - how does this assignment help that process or prevent early fcu?

Comment (tersec) [reply beacon_chain/consensus_object_pools/consensus_manager.nim]: `getProposalState`:
https://github.com/status-im/nimbus-eth2/blob/d3400ca11b165a313f6197722d274dcdafac3782/beacon_chain/consensus_object_pools/blockchain_dag.nim#L2299-L2330

It always uses `clearanceState`.

It does not care what's in `headState`.

Because it's an early fcU, and came in in the example logs I've been looking through via get block by root/request manager, 8 seconds into the slot, when it's fcU'd, and the consensus manager detect it should propose next slot, it does a proposal fcU.

This (Sepolia) block is selected to be the updated head, because it comes in via attestations, effectively. The sepolia blockcha.in confirms it was in the finalized chain as of a few hours ago.

To do so, the consensus manager fcU logic, therefore, needs to run the epoch transition to update withdrawal on this new state resulting from this new head block for which no opportunity appears to have arisen to place it in `clearanceState`.

```
[
  {
    "headBlockHash": "0xa29f67fb19c1a0dee35caf30ef1cf48e8cbfb3ddb1049fd9adf34d3dcf9d54ff",
    "safeBlockHash": "0x0655fe5e98d7e076023e9492d9f24219a45529cd14ff0795b9c0b56544359fc9",
    "finalizedBlockHash": "0x21acc74ea7273b9171a61c5044130bb71fb173e9ebfb2f8989cc0554ec639c0b"
  },
  {
    "timestamp": "0x64409D60",
    "prevRandao": "0x0517ff165294a5e4d0293d043f45853551bb15e23a7f63eae1115dc928c0d337",
    "suggestedFeeRecipient": "0x3826539cbd8d68dcf119e80b994557b4278cec9f",
    "withdrawals": [
      {
        "index": "0x511D0E",
        "validatorIndex": "0x3B2",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0xF04"
      },
      {
        "index": "0x511D0F",
        "validatorIndex": "0x3B5",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0xF04"
      },
      {
        "index": "0x511D10",
        "validatorIndex": "0x3B8",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0xF04"
      },
      {
        "index": "0x511D11",
        "validatorIndex": "0x3BB",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0xF04"
      },
      {
        "index": "0x511D12",
        "validatorIndex": "0x3BC",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0xF04"
      },
      {
        "index": "0x511D13",
        "validatorIndex": "0x3C0",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0xF04"
      },
      {
        "index": "0x511D14",
        "validatorIndex": "0x3C2",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0xF04"
      },
      {
        "index": "0x511D15",
        "validatorIndex": "0x3C6",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0xF04"
      },
      {
        "index": "0x511D16",
        "validatorIndex": "0x3C7",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0xF04"
      },
      {
        "index": "0x511D17",
        "validatorIndex": "0x3CB",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0xF04"
      },
      {
        "index": "0x511D18",
        "validatorIndex": "0x3CC",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0xF04"
      },
      {
        "index": "0x511D19",
        "validatorIndex": "0x3D5",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0xB43"
      },
      {
        "index": "0x511D1A",
        "validatorIndex": "0x3D8",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0xB43"
      },
      {
        "index": "0x511D1B",
        "validatorIndex": "0x3DA",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0xB43"
      },
      {
        "index": "0x511D1C",
        "validatorIndex": "0x3DD",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0x1DDBEF"
      },
      {
        "index": "0x511D1D",
        "validatorIndex": "0x3E0",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0xB43"
      }
    ]
  }
]
[
  {
    "headBlockHash": "0xa29f67fb19c1a0dee35caf30ef1cf48e8cbfb3ddb1049fd9adf34d3dcf9d54ff",
    "safeBlockHash": "0xa287cd3336ae0007aea9c7199d89c4dc97d205e5dd98ff7a952586ec003a74f5",
    "finalizedBlockHash": "0x21acc74ea7273b9171a61c5044130bb71fb173e9ebfb2f8989cc0554ec639c0b"
  },
  {
    "timestamp": "0x64409D60",
    "prevRandao": "0x0517ff165294a5e4d0293d043f45853551bb15e23a7f63eae1115dc928c0d337",
    "suggestedFeeRecipient": "0x3826539cbd8d68dcf119e80b994557b4278cec9f",
    "withdrawals": [
      {
        "index": "0x511D0E",
        "validatorIndex": "0x3AE",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0x2C027"
      },
      {
        "index": "0x511D0F",
        "validatorIndex": "0x3AF",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0x2C027"
      },
      {
        "index": "0x511D10",
        "validatorIndex": "0x3B0",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0x2C027"
      },
      {
        "index": "0x511D11",
        "validatorIndex": "0x3B1",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0x2C027"
      },
      {
        "index": "0x511D12",
        "validatorIndex": "0x3B2",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0x2CF2B"
      },
      {
        "index": "0x511D13",
        "validatorIndex": "0x3B3",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0x2C027"
      },
      {
        "index": "0x511D14",
        "validatorIndex": "0x3B4",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0x2C027"
      },
      {
        "index": "0x511D15",
        "validatorIndex": "0x3B5",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0x241F9"
      },
      {
        "index": "0x511D16",
        "validatorIndex": "0x3B6",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0x232F5"
      },
      {
        "index": "0x511D17",
        "validatorIndex": "0x3B7",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0x2C027"
      },
      {
        "index": "0x511D18",
        "validatorIndex": "0x3B8",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0x241F9"
      },
      {
        "index": "0x511D19",
        "validatorIndex": "0x3B9",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0x2C027"
      },
      {
        "index": "0x511D1A",
        "validatorIndex": "0x3BA",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0x2C027"
      },
      {
        "index": "0x511D1B",
        "validatorIndex": "0x3BB",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0x2CF2B"
      },
      {
        "index": "0x511D1C",
        "validatorIndex": "0x3BC",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0x241F9"
      },
      {
        "index": "0x511D1D",
        "validatorIndex": "0x3BD",
        "address": "0xe276bc378a527a8792b353cdca5b5e53263dfb9e",
        "amount": "0x2C027"
      }
    ]
  }
]
[
  {
    "headBlockHash": "0xb8fe4a6427c7ced962403de6baa9f7591110d88cd9d1435dba843ada1116b92e",
    "safeBlockHash": "0xa29f67fb19c1a0dee35caf30ef1cf48e8cbfb3ddb1049fd9adf34d3dcf9d54ff",
    "finalizedBlockHash": "0xa287cd3336ae0007aea9c7199d89c4dc97d205e5dd98ff7a952586ec003a74f5"
  },
  {
    "timestamp": "0x64409FA0",
    "prevRandao": "0xb41262284a4ec50f17621a97a42b40b1b9967df80d42b07d4c8be4a5ef381d95",
    "suggestedFeeRecipient": "0x3826539cbd8d68dcf119e80b994557b4278cec9f",
    "withdrawals": [
      {
        "index": "0x511FAE",
        "validatorIndex": "0x19A",
        "address": "0x25c4a76e7d118705e7ea2e9b7d8c59930d8acd3b",
        "amount": "0x294B"
      },
      {
        "index": "0x511FAF",
        "validatorIndex": "0x19D",
        "address": "0x25c4a76e7d118705e7ea2e9b7d8c59930d8acd3b",
        "amount": "0x294B"
      },
      {
        "index": "0x511FB0",
        "validatorIndex": "0x1A0",
        "address": "0x25c4a76e7d118705e7ea2e9b7d8c59930d8acd3b",
        "amount": "0x258A"
      },
      {
        "index": "0x511FB1",
        "validatorIndex": "0x1A5",
        "address": "0x25c4a76e7d118705e7ea2e9b7d8c59930d8acd3b",
        "amount": "0x258A"
      },
      {
        "index": "0x511FB2",
        "validatorIndex": "0x1A9",
        "address": "0x25c4a76e7d118705e7ea2e9b7d8c59930d8acd3b",
        "amount": "0x258A"
      },
      {
        "index": "0x511FB3",
        "validatorIndex": "0x1AE",
        "address": "0x25c4a76e7d118705e7ea2e9b7d8c59930d8acd3b",
        "amount": "0x21C9"
      },
      {
        "index": "0x511FB4",
        "validatorIndex": "0x1AF",
        "address": "0x25c4a76e7d118705e7ea2e9b7d8c59930d8acd3b",
        "amount": "0x21C9"
      },
      {
        "index": "0x511FB5",
        "validatorIndex": "0x1B0",
        "address": "0x25c4a76e7d118705e7ea2e9b7d8c59930d8acd3b",
        "amount": "0x21C9"
      },
      {
        "index": "0x511FB6",
        "validatorIndex": "0x1B1",
        "address": "0x25c4a76e7d118705e7ea2e9b7d8c59930d8acd3b",
        "amount": "0x21C9"
      },
      {
        "index": "0x511FB7",
        "validatorIndex": "0x1BC",
        "address": "0x25c4a76e7d118705e7ea2e9b7d8c59930d8acd3b",
        "amount": "0x21C9"
      },
      {
        "index": "0x511FB8",
        "validatorIndex": "0x1BE",
        "address": "0x25c4a76e7d118705e7ea2e9b7d8c59930d8acd3b",
        "amount": "0x1E08"
      },
      {
        "index": "0x511FB9",
        "validatorIndex": "0x1C6",
        "address": "0x25c4a76e7d118705e7ea2e9b7d8c59930d8acd3b",
        "amount": "0x1E08"
      },
      {
        "index": "0x511FBA",
        "validatorIndex": "0x1C7",
        "address": "0x25c4a76e7d118705e7ea2e9b7d8c59930d8acd3b",
        "amount": "0x1E08"
      },
      {
        "index": "0x511FBB",
        "validatorIndex": "0x1D4",
        "address": "0x25c4a76e7d118705e7ea2e9b7d8c59930d8acd3b",
        "amount": "0x1A47"
      },
      {
        "index": "0x511FBC",
        "validatorIndex": "0x1D7",
        "address": "0x25c4a76e7d118705e7ea2e9b7d8c59930d8acd3b",
        "amount": "0x1A47"
      },
      {
        "index": "0x511FBD",
        "validatorIndex": "0x1D9",
        "address": "0x25c4a76e7d118705e7ea2e9b7d8c59930d8acd3b",
        "amount": "0x1A47"
      }
    ]
  }
]
```
shows a couple different withdrawals sent by Nimbus. The first one is with an epoch state transition which appears to be from an incorrect previous state not incorporating the new head it had just received, and therefore does not match the second (`validator_duties` proposal) one.

Comment (arnetheduck) [reply beacon_chain/consensus_object_pools/consensus_manager.nim]: > It does not care what's in headState.

`getProposalState` calls `updateState` which copies the head state (or any other state) in order to arrive at the expected block/slot combo:

https://github.com/status-im/nimbus-eth2/blob/d3400ca11b165a313f6197722d274dcdafac3782/beacon_chain/consensus_object_pools/blockchain_dag.nim#L1433

There exists an inefficiency here: it always starts with `clearanceState` instead of first investigating which cached state is best and using that as basis for the initial copy.

Basically, the assignment you're making _should_ be a (logical) noop - if it's changing the logic of `getProposalState` in any way, there's a bug somewhere.

Comment (tersec) [reply beacon_chain/consensus_object_pools/consensus_manager.nim]: Okay. Will remove this, and at least just have it not fcU ahead in these circumstances. Bad blocks are better than no blocks (even if the engine API specifications say/should shortly say Nimbus is correct, and the ELs should account for the changes in withdrawals when creating payload IDs, there's a deployed base already).

Comment (tersec) [reply beacon_chain/consensus_object_pools/consensus_manager.nim]: Further investigation can wait for after this release.

Comment (arnetheduck) [reply beacon_chain/consensus_object_pools/consensus_manager.nim]: mmyeah - so "logically", what's returned by `getProposalState` should be a "pure" function of its given head/slot parameters - the fact that it copies `clearnanceState` is just an optimization / implementation detail, so any bugs in this area are likely to be derived from `head` or `slot` taking on unexpacted values (unless of course they're performance bugs, in which case, as noted, there exist "optimization opportunities" inside `getProposerState` itself.

Comment (tersec) [reply beacon_chain/consensus_object_pools/consensus_manager.nim]: https://github.com/status-im/nimbus-eth2/pull/4842/commits/48d8ae47d605443a8780fd5d6f948b120b7060c9

**Review by arnetheduck — APPROVED** [review]
